pyomo/core/tests/unit/test_indexed.py

```diff
@@ -200,6 +201,21 @@ def test_ordered_keys(self):
         self.assertEqual(list(zip(ordered_keys, [10, 20, 1, 30, 1])),
                          list(m.P.items(True)))
 
+    def test_index_attribute_out_of_sync(self):
+        m = ConcreteModel()
+        m.x = Var([1,2,3])
+        # make sure everything is right to begin with
+        for i in [1, 2, 3]:
+            self.assertEqual(m.x[i].index(), i)
+        # now mess it up
+        m.x[3]._index = 2
+        with self.assertRaisesRegex(
+                DeveloperError,
+                ".*The '_data' dictionary and '_index' attribute are out of "
+                "sync for index '2' on indexed component 'x': The '_data' "
+                "dictionary contains 'x\[2\]' as the value corresponding to "
+                "'2'"):
```

Comment:
I think this is correct, but it is a good point that the message is confusing.  @emma58, what do you think about a message more like:

```python
"The '_data' dictionary and '_index' attribute are out of sync for indexed "
"%s '%s': The %s entry in the '_data' dictionary does not map back to this "
"component data object' % (obj.ctype.__name__, obj.parent_component().name, obj._index) 
```

Reply:
Yes, @blnicho, they are both `2`'s because you would have to iterate over the `_data` dictionary to get the `3` (returning to the old inefficiency). But I like @jsiirola's suggestion--I'll make the change.